src/cudamatrix/cu-matrix.cc

```diff
@@ -968,6 +969,46 @@ void CuMatrixBase<Real>::AddMat(Real alpha, const CuMatrixBase<Real>& A,
 }
 
 template<typename Real>
+void CuMatrixBase<Real>::AddSmat(Real alpha, const CuSparseMatrix<Real> &A,
+                                 MatrixTransposeType trans) {
+#if HAVE_CUDA == 1
+  if (CuDevice::Instantiate().Enabled()) {
+    if (trans == kNoTrans) {
+      KALDI_ASSERT(NumRows() == A.NumRows());
+      KALDI_ASSERT(NumCols() == A.NumCols());
+    } else {
+      KALDI_ASSERT(NumRows() == A.NumCols());
+      KALDI_ASSERT(NumCols() == A.NumRows());
+    }
+
+    CuTimer tim;
+
+    // We use warpSize threads per row to access only the nonzero elements.
+    // Every CU1DBLOCK/warpSize rows share one thread block.
+    // 1D grid to cover all rows of A.
+    const int warpSize = 32;
```

Comment:
At this point in the code you seem to be using warpSize fixed to 32, but in the CUDA code you use just 'warpSize' which is not bound to this variable so as far as I can see might be dependent on the hardware.  Is this correct?